File: leetcode3/최원준/883. Projection Area of 3D Shapes.py
#

'''
1. 아이디어 :
xy는 1 이상의 grid[i][j]
xz는 각 row의 최대값의 합
yz는 각 col의 최대값의 

2. 시간복잡도 :
    O(n * n)

3. 자료구조/알고리즘 :
-

'''
class Solution:
    def projectionArea(self, grid: List[List[int]]) -> int:
        n = len(grid)

        """
        [ [1,2,3], [4,5,6], [7,8,9] ]
        """

        ans = 0
        for i in range(n):
            xz_max = 0
            yz_max = 0
            for j in range(n):
                ans += 1 if grid[i][j] != 0 else 0 # xy
                xz_max = max(xz_max, grid[i][j]) # xz
                yz_max = max(yz_max, grid[j][i]) # yz
            ans+=yz_max
            ans+=xz_max
                
        return ans

File: leetcode3/이진희/883_Projection Area of 3D Shapes.java
/* 

1. 아이디어 : 쌓여있는 블록의 3면의 합 구하기
              바닥은 블록 1개 이상 , row, col 기준 최댓값을을 각각 구해서 더해준다.

2. 시간복잡도 : O(N*M)*2

3. 자료구조/알고리즘 : 완전탐색

 */
class Solution {
    public int projectionArea(int[][] grid) {
        // 쌓여있는 모양
        // 가장자리 
        // 중간

        int ans = 0;
        for(int i=0; i<grid.length; i++) {
            int maxCol = 0;
            for(int j=0; j<grid[0].length; j++) {
                if(grid[i][j]>=1) ans++;
                maxCol = Math.max(maxCol, grid[i][j]);
            }
            ans+=maxCol;
        }

        for(int i=0; i<grid[0].length; i++) {
            int maxRow = 0;
            for(int j=0; j<grid.length; j++) {
                maxRow = Math.max(maxRow, grid[j][i]);
            }
            ans+=maxRow;
        }

        return ans;  
    }
}